packages/algorand-js/test/src/interpreter/opcode-list.ts

```diff
@@ -1118,49 +1190,186 @@ describe("Teal Opcodes", function () {
     });
   });
 
-  describe("substring3", function () {
+  describe("Substring", function () {
     const stack = new Stack<StackElem>();
+    const start = BigInt('0');
+    const end = BigInt('4');
 
     it("should return correct substring", function () {
-      stack.push(BigInt('4'));
-      stack.push(BigInt('0'));
       stack.push(toBytes("Algorand"));
-
-      const op = new Substring3();
+      const op = new Substring(start, end);
       op.execute(stack);
 
       const top = stack.pop() as Uint8Array;
       assert.equal("Algo", convertToString(top));
     });
 
-    it("should throw Invalid type error", function () {
+    it("should throw Invalid type error",
+      execExpectError(
+        stack,
+        [BigInt('1')],
+        new Substring(start, end),
+        ERRORS.TEAL.INVALID_TYPE
+      )
+    );
+
+    it("should throw error if start is not uint8", function () {
+      execExpectError(
+        stack,
+        [toBytes("Algorand")],
+        new Substring(BigInt(MIN_UINT8 - 5), end),
+        ERRORS.TEAL.INVALID_UINT8
+      );
+
+      execExpectError(
+        stack,
+        [toBytes("Algorand")],
+        new Substring(BigInt(MAX_UINT8 + 5), end),
+        ERRORS.TEAL.INVALID_UINT8
+      );
+    });
+
+    it("should throw error if end is not uint8", function () {
+      execExpectError(
+        stack,
+        [toBytes("Algorand")],
+        new Substring(start, BigInt(MIN_UINT8 - 5)),
+        ERRORS.TEAL.INVALID_UINT8
+      );
+
+      execExpectError(
+        stack,
+        [toBytes("Algorand")],
+        new Substring(start, BigInt(MAX_UINT8 + 5)),
+        ERRORS.TEAL.INVALID_UINT8
+      );
+    });
+
+    it("should throw error because start > end",
+      execExpectError(
+        stack,
+        [toBytes("Algorand")],
+        new Substring(end + BigInt('5'), end),
+        ERRORS.TEAL.SUBSTRING_END_BEFORE_START
+      )
+    );
+
+    it("should throw error because range beyong string",
+      execExpectError(
+        stack,
+        [toBytes("Algorand")],
+        new Substring(start, start + BigInt('40')),
+        ERRORS.TEAL.SUBSTRING_RANGE_BEYOND
+      )
+    );
+  });
+
+  describe("Substring3", function () {
+    const stack = new Stack<StackElem>();
+
+    it("should return correct substring", function () {
       stack.push(BigInt('4'));
       stack.push(BigInt('0'));
-      stack.push(BigInt('1234'));
+      stack.push(toBytes("Algorand"));
 
       const op = new Substring3();
-      expectTealError(
-        () => op.execute(stack),
+      op.execute(stack);
+
+      const top = stack.pop() as Uint8Array;
```

Comment:
are we supposed to use `as Uint8Array` in tests ?

Reply:
Yes. The below functions takes `Uint8Array` as an argument. Hence the `as Uint8Array`. Do you suggest something else?

Reply:
we can convert "Algo" to bytes